### ISSUE TITLE
Devtools: infinite recursive loop in devtools driver

### ISSUE THREAD

**Issue by dylang** [webdriverio/webdriverio#5048]
**Environment:**
 - **WebdriverIO version:** webdriverio@5.18.7
 - **Mode:** standalone
 - **Node.js version:** 10.16.0
 - **yarn version:** 1.22.0
 - **Browser name and version:** 80
 - **Platform name and version:** MacOS 10.14
 - **Additional wdio packages used (if applicable):** devtools

**Config of WebdriverIO**
`browserName: chrome`

**Describe the bug**
https://github.com/webdriverio/webdriverio/blob/master/packages/devtools/src/devtoolsdriver.js#L197

if `pageloadTimeoutReached` is falsy, it will keep calling `this. checkPendingNavigations` until memory runs out:

**To Reproduce**

Hitting control-c sometimes while tests are running can trigger this.

Running node with `node --prof --expose-gc --inspect` makes it easy to see in Chrome the infinite loop.

<img width="248" alt="Screen Shot 2020-02-20 at 4 37 06 PM" src="https://user-images.githubusercontent.com/51505/74980891-46ee2700-53ff-11ea-9f3b-086036bb40bc.png">

**Expected behavior**

* Don't use recursion when we don't know how deep it could get. Maybe [p-retry](https://github.com/sindresorhus/p-retry)?
* With control-c, it should gracefully shut down the devtools package.  Something like this maybe?
```ts
// If the parent of the spawned process becomes 1, then it means the process that spawned our process was killed and we should exit too.
if (process.ppid === 1) {
 process.exit();
}
```

**Log**

```
[88676:0x102802000]    96774 ms: Scavenge 1378.0 (1424.6) -> 1374.3 (1426.1) MB, 2.8 / 0.0 ms  (average mu = 0.330, current mu = 0.315) allocation failure
[88676:0x102802000]    96782 ms: Scavenge 1378.2 (1426.1) -> 1374.5 (1427.1) MB, 2.3 / 0.0 ms  (average mu = 0.330, current mu = 0.315) allocation failure
[88676:0x102802000]    96790 ms: Scavenge 1378.3 (1427.1) -> 1374.6 (1428.6) MB, 2.9 / 0.0 ms  (average mu = 0.330, current mu = 0.315) allocation failure


<--- JS stacktrace --->

==== JS stack trace =========================================

    0: ExitFrame [pc: 0x3c22f775be3d]
    1: StubFrame [pc: 0x3c22f775d3c6]
Security context: 0x1e9678a9e6e9 <JSObject>
    2: CallSiteToString [0x1e9622f63b09] [/Users/dgreene/medallia/perf-report/node_modules/source-map-support/source-map-support.js:~262] [pc=0x3c22f7e841e5](this=0x1e965b225411 <Object map = 0x1e96113352c1>)
    3: StubFrame [pc: 0x3c22f7713e51]
    4: StubFrame [pc: 0x3c22f775e15b]
    5: prepareStackTrace [0x1e96ba71a...

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x10003cf99 node::Abort() [/Users/dgreene/.nvm/versions/node/v10.16.0/bin/node]
 2: 0x10003d1a3 node::OnFatalError(char const*, char const*) [/Users/dgreene/.nvm/versions/node/v10.16.0/bin/node]
 3: 0x1001b7835 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/Users/dgreene/.nvm/versions/node/v10.16.0/bin/node]
 4: 0x100585682 v8::internal::Heap::FatalProcessOutOfMemory(char const*) [/Users/dgreene/.nvm/versions/node/v10.16.0/bin/node]
 5: 0x100588155 v8::internal::Heap::CheckIneffectiveMarkCompact(unsigned long, double) [/Users/dgreene/.nvm/versions/node/v10.16.0/bin/node]
 6: 0x100583fff v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/Users/dgreene/.nvm/versions/node/v10.16.0/bin/node]
 7: 0x1005821d4 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/Users/dgreene/.nvm/versions/node/v10.16.0/bin/node]
 8: 0x10058ea6c v8::internal::Heap::AllocateRawWithLigthRetry(int, v8::internal::AllocationSpace, v8::internal::AllocationAlignment) [/Users/dgreene/.nvm/versions/node/v10.16.0/bin/node]
 9: 0x10058eaef v8::internal::Heap::AllocateRawWithRetryOrFail(int, v8::internal::AllocationSpace, v8::internal::AllocationAlignment) [/Users/dgreene/.nvm/versions/node/v10.16.0/bin/node]
10: 0x10055e434 v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationSpace) [/Users/dgreene/.nvm/versions/node/v10.16.0/bin/node]
11: 0x1007e6714 v8::internal::Runtime_AllocateInNewSpace(int, v8::internal::Object**, v8::internal::Isolate*) [/Users/dgreene/.nvm/versions/node/v10.16.0/bin/node]
12: 0x3c22f775be3d

<--- Last few GCs --->

[88749:0x102802000]    96322 ms: Scavenge 1382.9 (1424.1) -> 1379.2 (1426.1) MB, 3.3 / 0.0 ms  (average mu = 0.307, current mu = 0.269) allocation failure
[88749:0x102802000]    96333 ms: Scavenge 1383.0 (1426.1) -> 1379.3 (1427.6) MB, 3.2 / 0.0 ms  (average mu = 0.307, current mu = 0.269) allocation failure
[88749:0x102802000]    96344 ms: Scavenge 1383.2 (1427.6) -> 1379.5 (1429.6) MB, 3.4 / 0.0 ms  (average mu = 0.307, current mu = 0.269) allocation failure


<--- JS stacktrace --->

==== JS stack trace =========================================

    0: ExitFrame [pc: 0x3f86fc5be3d]
    1: StubFrame [pc: 0x3f86fc41b83]
    2: StubFrame [pc: 0x3f86fc1f79e]
    3: EntryFrame [pc: 0x3f86fc04ba1]
    4: ExitFrame [pc: 0x3f86fc5be3d]
Security context: 0x1cf46ff1e6e9 <JSObject>
    5: _tickCallback [0x1cf481f866f9] [internal/process/next_tick.js:68] [bytecode=0x1cf473468f11 offset=157](this=0x1cf4d1483c81 <process map = 0x1cf4ec466389>)
    6: InternalFrame [pc: 0x3f86fc0ee75]
    7: En...

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x10003cf99 node::Abort() [/Users/dgreene/.nvm/versions/node/v10.16.0/bin/node]
 2: 0x10003d1a3 node::OnFatalError(char const*, char const*) [/Users/dgreene/.nvm/versions/node/v10.16.0/bin/node]
 3: 0x1001b7835 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/Users/dgreene/.nvm/versions/node/v10.16.0/bin/node]
 4: 0x100585682 v8::internal::Heap::FatalProcessOutOfMemory(char const*) [/Users/dgreene/.nvm/versions/node/v10.16.0/bin/node]
 5: 0x100588155 v8::internal::Heap::CheckIneffectiveMarkCompact(unsigned long, double) [/Users/dgreene/.nvm/versions/node/v10.16.0/bin/node]
 6: 0x100583fff v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/Users/dgreene/.nvm/versions/node/v10.16.0/bin/node]
 7: 0x1005821d4 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/Users/dgreene/.nvm/versions/node/v10.16.0/bin/node]
 8: 0x10058ea6c v8::internal::Heap::AllocateRawWithLigthRetry(int, v8::internal::AllocationSpace, v8::internal::AllocationAlignment) [/Users/dgreene/.nvm/versions/node/v10.16.0/bin/node]
 9: 0x10058eaef v8::internal::Heap::AllocateRawWithRetryOrFail(int, v8::internal::AllocationSpace, v8::internal::AllocationAlignment) [/Users/dgreene/.nvm/versions/node/v10.16.0/bin/node]
10: 0x10055e434 v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationSpace) [/Users/dgreene/.nvm/versions/node/v10.16.0/bin/node]
11: 0x1007e6714 v8::internal::Runtime_AllocateInNewSpace(int, v8::internal::Object**, v8::internal::Isolate*) [/Users/dgreene/.nvm/versions/node/v10.16.0/bin/node]
12: 0x3f86fc5be3d
````

**Comment by christian-bromann** [comment]
Can you provide a reproducible example?

**Comment by dylang** [comment]
Sure thing, I'll have one in a few minutes.

**Comment by pmerwin** [comment]
FYI: We are seeing a similar running chromium 80 in our lambda, I can't reproduce it locally, but I may be able to provide the logs from the crash in our lambda.

**Comment by dylang** [comment]
Simple repo case:
https://github.com/dylang/webdriverio-issue-repo

**Comment by dylang** [comment]
My example does not use `@wdio-cli` - but it still can happen with it, it's just much less likely:

I added the following code to `devtoolsdriver.js`:

```js
    console.log('devtools checkPendingNavigations pid', process.pid, 'ppid', process.ppid);
```

Ran `yarn wdio wdio.conf.js`

and hit `control-c` once.

Each time `devtools checkPendingNavigations` is printed is another recursive loop, which will eventually run out of memory if the parent process doesn't kill it soon enough.

```
^C

Ending WebDriver sessions gracefully ...
(press ctrl+c again to hard kill the runner)
2020-02-24T17:20:45.478Z INFO @wdio/local-runner: Shutting down spawned worker

[0-0] devtools checkPendingNavigations pid 33419 ppid 33417
[0-0] 2020-02-24T17:20:45.496Z INFO @wdio/local-runner: Run worker command: endSession
[0-0] devtools checkPendingNavigations pid 33419 ppid 33417
[0-0] 2020-02-24T17:20:45.497Z INFO @wdio/local-runner: Received SIGINT, giving process 5000ms to shutdown gracefully
[0-0] devtools checkPendingNavigations pid 33419 ppid 33417
[0-0] devtools checkPendingNavigations pid 33419 ppid 33417
[0-0] devtools checkPendingNavigations pid 33419 ppid 33417
[0-0] devtools checkPendingNavigations pid 33419 ppid 33417
[0-0] devtools checkPendingNavigations pid 33419 ppid 33417
[0-0] devtools checkPendingNavigations pid 33419 ppid 33417
[0-0] devtools checkPendingNavigations pid 33419 ppid 33417
[0-0] devtools checkPendingNavigations pid 33419 ppid 33417
[0-0] devtools checkPendingNavigations pid 33419 ppid 33417
[0-0] devtools checkPendingNavigations pid 33419 ppid 33417
[0-0] devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
[0-0] devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
[0-0] devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
devtools checkPendingNavigations pid 33419 ppid 33417
```

**Comment by dylang** [comment]
FWIW, this change fixed the issue for me:

Change this line:
https://github.com/webdriverio/webdriverio/blob/master/packages/devtools/src/devtoolsdriver.js#L197

to this:

```js
      return new Promise((resolve) => {
        setTimeout(() => resolve(this.checkPendingNavigations(pendingNavigationStart)), 1);
      });
```

**Comment by dylang** [comment]
There's possibly a better non-recursive solution with a module like [p-retry](https://github.com/sindresorhus/p-retry), but this at least allows the process to exit instead of being stuck in a recursive loop until the heap is used up.

**Comment by dylang** [comment]
Hi @christian-bromann, would you like a PR for either of these changes? It would be extremely helpful for us to get this in a v5 update.

**Comment by christian-bromann** [comment]
> would you like a PR for either of these changes?

@dylang yes please!

**Comment by christian-bromann** [comment]
This has been addressed in #5404. Let's continue our conversation there.
